### ISSUE TITLE
3.0.12 Datasource: Segmentation fault: heavy use datasource?

### ISSUE THREAD

**Issue by jc-vogel** [cakephp/cakephp#7299]
PHP throws a Segmentation fault when I use the database quite a lot. I have a Shell that calls the custom finder below around 1500 times, sometimes it throws this segmentation fault, sometimes it doesn't. It's not related to the entity of query itself, when I change the code a bit (add echo statement) it happens at another place. But as far as I can see it's not the memory usage 
When it's thrown, it's always from the custom finder below. I think it has something to do with heavy usage of the database Connection. 
I'm trying to debug this, but I'm getting a bit stuck. 

I have the following custom finder:

```
public function findFull(Query $query, array $options)
    {
        $query->contain([
            'Addresses',
            'Contracts',
            'Ovchipcards.Students.Users',
            'PaymentMethods',
            'Routes.RouteDepartures',
            'Routes.RouteDestinations',
            'Routes.RoutePrices',
            'Schoollocations.Addresses',
            'Schoollocations.Schools.SchoolBlacklists',
            'Statuses' => function ($q) {
                return $q
                    ->order(['SubscriptionStatuses.created' => 'DESC']);
            },
        ])
        ->where([
            'Subscriptions.id' => $options['id']
        ]);

        return $query;
    }
```

I created a shell to try to find the issue:
The following code will always crash with a segmentation fault after 15 calls to findFull(): 

```
public function main()
    {
        parent::main();

        $subscriptionstable = TableRegistry::get('Subscriptions');
        // get all subscriptions from this table:
        $subscriptions = $subscriptionstable->find();
        // loop trough these subscriptions and find all contained data:
        foreach ($subscriptions as $subscription) {
            for($i = 0; $i < 100; $i++) {
                $full = $subscriptionstable->find('full', ['id' => $subscription->id])->first();
            }
            if (empty($full)) {
                continue;
            }
            echo $full->ident. '-'. memory_get_peak_usage().PHP_EOL;
        }

    }
```

Can anyone help me with this? Resetting the connection with $conn->disconnect() and $conn->connect() doesn't help.

**Comment by lorenzo** [comment]
I think it might be good to report this as well to the PHP team, since this kind of errors is a problem in the language itself. It would also help to know what version of PHP and database you are using :)

**Comment by vaughany** [comment]
I'm doing some development and after a while I'll start receiving segfaults also:

```
[Wed Sep 30 21:26:33.581655 2015] [core:notice] [pid 1701] AH00052: child pid 8499 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581702 2015] [core:notice] [pid 1701] AH00052: child pid 8833 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581710 2015] [core:notice] [pid 1701] AH00052: child pid 8853 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581719 2015] [core:notice] [pid 1701] AH00052: child pid 9150 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581726 2015] [core:notice] [pid 1701] AH00052: child pid 9201 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581735 2015] [core:notice] [pid 1701] AH00052: child pid 9207 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:33.581740 2015] [core:notice] [pid 1701] AH00052: child pid 9941 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:39.590090 2015] [core:notice] [pid 1701] AH00052: child pid 9534 exit signal Segmentation fault (11)
zend_mm_heap corrupted
zend_mm_heap corrupted
[Wed Sep 30 21:26:54.607286 2015] [core:notice] [pid 1701] AH00052: child pid 9943 exit signal Segmentation fault (11)
[Wed Sep 30 21:26:59.613403 2015] [core:notice] [pid 1701] AH00052: child pid 9944 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:16.631611 2015] [core:notice] [pid 1701] AH00052: child pid 7874 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:16.631654 2015] [core:notice] [pid 1701] AH00052: child pid 9945 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:16.631659 2015] [core:notice] [pid 1701] AH00052: child pid 9946 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:16.631665 2015] [core:notice] [pid 1701] AH00052: child pid 9947 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:21.640566 2015] [core:notice] [pid 1701] AH00052: child pid 9949 exit signal Segmentation fault (11)
[Wed Sep 30 21:27:51.670038 2015] [core:notice] [pid 1701] AH00052: child pid 9950 exit signal Segmentation fault (11)
[Wed Sep 30 21:33:59.989660 2015] [mpm_prefork:notice] [pid 10166] AH00163: Apache/2.4.10 (Ubuntu) configured -- resuming normal operations
[Wed Sep 30 21:33:59.989694 2015] [core:notice] [pid 10166] AH00094: Command line: '/usr/sbin/apache2'
[Wed Sep 30 21:34:58.049639 2015] [core:notice] [pid 10166] AH00052: child pid 10170 exit signal Segmentation fault (11)
[Wed Sep 30 21:34:58.049669 2015] [core:notice] [pid 10166] AH00052: child pid 10171 exit signal Segmentation fault (11)
[Wed Sep 30 21:34:58.049674 2015] [core:notice] [pid 10166] AH00052: child pid 10172 exit signal Segmentation fault (11)
[Wed Sep 30 21:34:58.049679 2015] [core:notice] [pid 10166] AH00052: child pid 10173 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:00.055858 2015] [core:notice] [pid 10166] AH00052: child pid 10169 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:00.055889 2015] [core:notice] [pid 10166] AH00052: child pid 10177 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:00.055897 2015] [core:notice] [pid 10166] AH00052: child pid 10179 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:03.633827 2015] [mpm_prefork:notice] [pid 10166] AH00169: caught SIGTERM, shutting down
[Wed Sep 30 21:35:04.711707 2015] [mpm_prefork:notice] [pid 10233] AH00163: Apache/2.4.10 (Ubuntu) configured -- resuming normal operations
[Wed Sep 30 21:35:04.711743 2015] [core:notice] [pid 10233] AH00094: Command line: '/usr/sbin/apache2'
[Wed Sep 30 21:35:24.732572 2015] [core:notice] [pid 10233] AH00052: child pid 10237 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:24.732603 2015] [core:notice] [pid 10233] AH00052: child pid 10238 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:24.732608 2015] [core:notice] [pid 10233] AH00052: child pid 10239 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:24.732613 2015] [core:notice] [pid 10233] AH00052: child pid 10240 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:29.739876 2015] [core:notice] [pid 10233] AH00052: child pid 10244 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:33.747109 2015] [core:notice] [pid 10233] AH00052: child pid 10236 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:33.747172 2015] [core:notice] [pid 10233] AH00052: child pid 10247 exit signal Segmentation fault (11)
[Wed Sep 30 21:35:33.747180 2015] [core:notice] [pid 10233] AH00052: child pid 10248 exit signal Segmentation fault (11)
```

As you can see I'm restarting Apache and getting segfaults less than a minute later.  I don't know how to debug this to be honest, but as I found this bug report I thought I should add my 2p also.  Let me know if there's anything more I can do to report more detail here, if needs be.

I'm using CakePHP 3.1 and PHP 5.6.4 on Ubuntu 15.04.

**Comment by lorenzo** [comment]
Getting a way to reproduce the issue would be great

**Comment by jc-vogel** [comment]
I reproduced it by brute-forcing a customer finder, see above. I don't have much time for now, this weekend I'll try to create a failing test

**Comment by lorenzo** [comment]
@jc-vogel PHP version?

**Comment by vaughany** [comment]
I am unable to consistently reproduce the issue, unfortunately.

It seems to happen when I make a mistake with the code (I'm new to CakePHP and occasionally make daft mistakes) resulting in a lovely stack trace, but then correcting the error and refreshing my browser causes a 'no data returned' error in the browser and the lines in the Apache log as seen above.

Could a day's constant developing and browser refreshing result in memory fragmentation and Apache/PHP to fail?  I don't use the computer for much else.  It's also running MySQL (or MariaDB), as well as a few more tabs in Chrome (e.g. Gmail), a Bash or two (one running tail -f  on the Apache error log), Sublime Text and sometimes a music player.

If/when it happens again I'll do what I can to log what I've been doing and what's running.

**Comment by jc-vogel** [comment]
@lorenzo 
php: 5.6.13
cakephp: 3.0.13

A day's constant developing is not the problem I guess. I have the problem after a fresh restart.

**Comment by lorenzo** [comment]
@jc-vogel are you still experiencing this issue?

**Comment by lorenzo** [comment]
Closing as we cannot find a way of reproducing this issue

**Comment by half2me** [comment]
Could it have something to do with this issue?
https://github.com/cakephp/cakephp/issues/8162

**Comment by jippi** [comment]
No :)

**Comment by sanjulikar** [comment]
Hey I am facing the same issue 'child pid 11255 exit signal Segmentation fault (11)' when trying to print value from database/Auth data, meanwhile using paginator also(I know using pr() is a bad habit but I have been using it for so long and never faced issue with it).
pr() die is going fine when displaying a non-database data.
versions:
Cakephp: 3.2.5
PHP 5.6.4-4
Apache : Apache/2.4.10
Ubuntu:15.04
Os type : 32 bit
ps: I have been coding from many years but I just upgraded to Cakephp 3
Code: 
(Very basic Index cake bake code)

public function index() {
        $users = $this->paginate($this->Users);
        pr($this->Auth->user()); 
// if I use $this->log($this->Auth->user()); , then there is no issue, issue is same for pr($users);
        $this->set(compact('users'));
        $this->set('_serialize', ['users']);
    }


Please help as I am heading a very imp project.

Thanks in advance.

**Comment by thinkingmedia** [comment]
@sanjulikar Ubuntu 15.04 has disconnected support. You should upgrade to 15.10. Not that this will fix your problem but the repos are not building PHP updates for 15.04 any more.

**Comment by chinpei215** [comment]
I could reproduce this issue. Also, I confirmed that if I called the [gc_collect_cycles()](http://php.net/manual/en/function.gc-collect-cycles.php) function explicitly at the beginning of the foreach loop, it eased the problem. I hope my comment helps someone who encountered this issue.

Lastly, I would like to thank **icchii**, a member of our Slack team, for his idea of using the `gc_collect_cycles` function. Thank you.
